src/gc.rs

```diff
@@ -0,0 +1,257 @@
+                    objptr as usize,
+                    nzptr.get(),
+                ));
+                flzr.finalize()
+            }
+            None => return,
+        }
+    }
+
+    fn new_from_layout(layout: Layout) -> NonNull<GcBox<MaybeUninit<T>>> {
+        unsafe {
+            let (nl, _) = Layout::new::<usize>().extend(layout).unwrap();
+            let base_ptr = GC_ALLOCATOR.alloc(nl).unwrap().0.as_ptr() as *mut usize;
+
+            // Placeholder for vptr
+            ::std::ptr::write(base_ptr as *mut Option<NonZeroUsize>, None);
+
+            NonNull::new_unchecked((base_ptr.add(1)) as *mut GcBox<MaybeUninit<T>>)
+        }
+    }
+}
+
+impl<T> GcBox<MaybeUninit<T>> {
+    unsafe fn assume_init(&mut self) -> NonNull<GcBox<T>> {
+        // With T now considered initialized, we must make sure that if GcBox<T>
+        // is reclaimed, T will be dropped. We need to find its vptr and replace the
+        // GcDummyDrop vptr in the block header with it.
+        let fatptr: &mut dyn Finalizable = self;
+        let vptr = NonZeroUsize::new_unchecked(
+            transmute::<&mut dyn Finalizable, (usize, usize)>(fatptr).1,
+        );
+        let base_ptr = (self as *mut _ as *mut usize).sub(1) as *mut Option<NonZeroUsize>;
+        let flzr = Self::finalizer_shim;
+
+        ::std::ptr::write(base_ptr, Some(vptr));
+        boehm::GC_register_finalizer(
+            base_ptr as *mut _ as *mut ::std::ffi::c_void,
+            flzr,
+            ::std::ptr::null_mut(),
+            ::std::ptr::null_mut(),
+            ::std::ptr::null_mut(),
+        );
+        NonNull::new_unchecked(self as *mut _ as *mut GcBox<T>)
+    }
+}
+
+/// Used to clean up resources after a garbage collection.
+///
+/// A `Finalizable` trait can be thought of as similar to `Drop`. Its `finalize`
+/// method is called by the GC when the object is considered garbage. We avoid
+/// `Drop` and use an explicit `Finalizable` trait for this because `Drop` is
+/// special-cased in compiler and we want to avoid prematurely dropping fields
+/// pointing to other `GcBox`s.
+trait Finalizable {
+    fn finalize(&mut self);
+}
+
+impl<T> Finalizable for GcBox<T> {
+    fn finalize(&mut self) {
+        unsafe { ManuallyDrop::drop(&mut self.0) };
+    }
+}
+
+impl<T: ?Sized> Deref for Gc<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*(self.ptr.as_ptr() as *const T) }
+    }
+}
+
+impl<T: ?Sized> DerefMut for Gc<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { &mut *(self.ptr.as_ptr() as *mut T) }
+    }
+}
+
+/// `Copy` and `Clone` are implemented manually because a reference to `Gc<T>`
+/// should be copyable regardless of `T`. It differs subtly from `#[derive(Copy,
+/// Clone)]` in that the latter only makes `Gc<T>` copyable if `T` is.
+impl<T: ?Sized> Copy for Gc<T> {}
+
+impl<T: ?Sized> Clone for Gc<T> {
+    fn clone(&self) -> Self {
+        Gc::from_inner(self.ptr)
```

Comment:
Can't we say something simpler like `*self` to show that this is just `Copy` in disguise?

Reply:
Okay

Reply:
Fixed: c0106ab90e2cf9f0e165b3e28c41dbc838aacd93